api/s2n.h

```diff
@@ -1973,6 +1973,34 @@ S2N_API extern int s2n_connection_free(struct s2n_connection *conn);
  */
 S2N_API extern int s2n_shutdown(struct s2n_connection *conn, s2n_blocked_status *blocked);
 
+/**
+ * Attempts to close the write side of the TLS connection.
+ *
+ * TLS1.3 supports closing the write side of a TLS connection while leaving the read
+ * side unaffected. This feature is usually referred to as "half-close". We send
+ * a close_notify alert, but do not wait for the peer to respond.
+ *
+ * Like `s2n_shutdown()`, this method does not affect the underlying transport.
+ *
+ * `s2n_shutdown_write()` may still be called for earlier TLS versions, but most
+ * TLS implementations will react by immediately discarding any pending writes and
+ * closing the connection.
+ *
+ * Once `s2n_shutdown_write()` is complete:
+ * * The s2n_connection handle CANNOT be used for writing.
+ * * The s2n_connection handle CAN be used for reading.
+ * * The write side of the underlying transport can be closed. Most likely via `shutdown()`.
+ *
+ * The application should still call `s2n_shutdown()` or wait for `s2n_recv()` to
+ * return 0 to indicate end-of-data before cleaning up the connection or closing
+ * the read side of the underlying transport.
+ *
+ * @param conn A pointer to the s2n_connection object
+ * @param blocked A pointer which will be set to the blocked status, as in s2n_negotiate()
+ * @returns S2N_SUCCESS on success. S2N_FAILURE on failure
+ */
+S2N_API extern int s2n_shutdown_write(struct s2n_connection *conn, s2n_blocked_status *blocked);
```

Comment:
What about s2n_half_close()?

Reply:
I considered that, but I'm worried about the shutdown vs close terminology change, and that "half close" doesn't obviously translate to "close the write side".

The POSIX way to do this is to call [shutdown](https://www.man7.org/linux/man-pages/man2/shutdown.2.html) with an argument specifying the direction, so `shutdown(fd, SHUT_WR)`. I can't add an argument to s2n_shutdown, but maybe I could make "s2n_shutdown_half(conn, how)" or "s2n_shutdown_partial(conn, how)" instead of s2n_shutdown_write?

Reply:
But I'm a little loathe to do that because closing the read side of a TLS connection isn't really a thing. You can't communicate to your peer that you've done that. It'd just mark it closed locally.

Reply:
One thought on the naming: do we want to use `write` even though the API is called `send`? Conceptually they're the same, but it would be nice to be consistent and indicate that they're connected.

Reply:
That's a very good point. I'll swap "write" for "send".